plugins/oauth/server/providers/base.py

```diff
@@ -0,0 +1,173 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+import json
+import re
+import requests
+import six
+
+from girder.api.rest import RestException
+from girder.constants import SettingKey
+from girder.utility import config, model_importer
+
+
+class ProviderBase(model_importer.ModelImporter):
+    def __init__(self, redirectUri, clientId=None, clientSecret=None):
+        """
+        Base class for OAuth providers. The purpose of these classes is to
+        perform the user information lookup to their respective provider
+        services in order to return a Girder user given an OAuth code that was
+        sent to us via a callback.
+
+        :param clientId: The client ID for the given OAuth provider.
+        :type clientId: str
+        :param clientSecret: The client secret for the given OAuth provider.
+        :type clientSecret: str
+        :param redirectUri: The redirect URI used in this OAuth flow.
+        :type redirectUri: str
+        """
+        self.clientId = clientId or self.getClientIdSetting()
+        self.clientSecret = clientSecret or self.getClientSecretSetting()
+        self.redirectUri = redirectUri
+
+    def getClientIdSetting(self):
+        raise NotImplementedError()
+
+    def getClientSecretSetting(self):
+        raise NotImplementedError()
+
+    @classmethod
+    def getUrl(cls, state):
+        raise NotImplementedError()
+
+    def getJson(self, **kwargs):
+        """
+        Make an HTTP request using the specified kwargs, then parse it as JSON
+        and return the value. If an error occurs, this raises an appropriate
+        exception containing the information.
+        """
+        resp = requests.request(**kwargs)
+        content = resp.content
+
+        if not isinstance(content, six.text_type):
+            content = content.decode('utf8')
+
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError:
+            raise RestException(
+                'Got %s from %s, response="%s".' % (
+                    resp.status_code, kwargs['url'], content
+                ), code=502)
+
+        try:
+            return json.loads(content)
+        except ValueError:
+            raise RestException('Non-JSON response: %s' % content, code=502)
+
+    def getUser(self, code):
+        """
+        Returns a user via this OAuth provider given an OAuth code. Subclasses
+        must implement this to retrieve an email address, first name, last name,
+        and optionally a login name from the OAuth provider, and then call
+        self.createOrReuseUser with that information to lookup or create a
+        corresponding Girder user.
+        """
+        raise NotImplementedError()
+
+    def createOrReuseUser(self, email, firstName, lastName, userName=None):
+        cursor = self.model('user').find({'email': email}, limit=1)
```

Comment:
@zachmullen One last question, why are we doing this ``cursor.count`` stuff instead of ``findOne``?